### PR TITLE
ui-manchette: refactor manchette props

### LINKED ISSUE #645: ui-manchette: redesign props
### Description and goal

Current props are very OSRD-specific: https://github.com/OpenRailAssociation/osrd-ui/blob/dev/ui-manchette/src/types.ts

Proposal:

```ts
type Waypoint = {
  id: string
  name: string
  secondaryCode: string
  position: number // en mm
}

type TrainPath = {
  id: number
  name: string
  points: { time: number; position: number }[][] // à re-discuter avec Alexis et Mathieu
}

type SpaceTimeChartWithManchetteProps = {
  paths: TrainPath[]
  selectedTrainId: number
  waypoints: Waypoint[]
}

type ManchetteProps = {
  waypoints: Waypoint[];
}
```

### Acceptance criteria

.

### PR DESCRIPTION
**Pull request by clarani**
closes #645

### CODE REVIEW

**Comment by clarani** [comment]
Draft until https://github.com/OpenRailAssociation/osrd-ui/pull/649 is merged

File: ui-manchette/src/types.ts [hunk 1]
```diff
@@ -1,73 +1,24 @@
...
-  track: string;
-};
-
-export type ArrayElement<ArrayType extends readonly unknown[] | null | undefined> =
-  ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
-
-export type OperationalPointType = ArrayElement<PathProperties['operational_points'] | null>;
-
-export type StyledOperationalPointType = OperationalPointType & {
+export type EnrichedWaypoint = Waypoint & {
```
Comment (emersion) [inline ui-manchette/src/types.ts]: Mandatory naming nitpicking: "enriched" is a bit vague. Would `InteractiveWaypoint` be better?

File: ui-manchette/src/types.ts [hunk 2]
```diff
@@ -1,73 +1,24 @@
...
-    | {
-        extensions?: OperationalPointExtensions;
-        id: string;
-        part: OperationalPointPart;
-        /** Distance from the beginning of the path in mm */
-        position: number;
-      }[]
-    | null;
+export type Waypoint = {
+  id: string;
```
Comment (emersion) [inline ui-manchette/src/types.ts:4]: Before this patch, there was no unique waypoint ID passed in. The OP ID was passed in, but the same OP could appear multiple times in the path.

I think a unique waypoint ID is a great idea, see https://github.com/OpenRailAssociation/osrd-ui/pull/668#discussion_r1816589010

Which flavor is this field? If it needs to be unique, would be great to document!

Comment (clarani) [reply ui-manchette/src/types.ts:4]: Yes, id here is the id of the waypoint (= `pathStep.id`), and not the OP ID

File: ui-manchette-with-spacetimechart/src/hooks/usePaths.ts [hunk 3]
```diff
@@ -1,26 +1,23 @@
...
-  type ProjectPathTrainResult,
-  type SpaceTimeCurves,
-} from '@osrd-project/ui-manchette/dist/types';
+import { type ProjectPathTrainResult } from '@osrd-project/ui-manchette/dist/types';
 
-const transformCurve = (curve: SpaceTimeCurves, departureTime: string) =>
+const transformCurve = (curve: ProjectPathTrainResult['spaceTimeCurves'][0], departureTime: Date) =>
   curve.positions.map((position, i) => ({
-    time: curve.times[i] + new Date(departureTime).getTime(),
+    time: curve.times[i] + departureTime.getTime(),
```
Comment (emersion) [inline ui-manchette-with-spacetimechart/src/hooks/usePaths.ts:8]: Oh nice, we're no longer parsing a date string at each curve point!

File: ui-manchette/src/components/OperationalPointList.tsx [hunk 4]
```diff
@@ -1,25 +1,22 @@
...
-const OperationalPointList = ({
-  operationalPoints,
-  activeOperationalPointId,
-}: OperationalPointListProps) => (
+const OperationalPointList = ({ waypoints, activeWaypointId }: OperationalPointListProps) => (
   <div className="operational-point-list ">
-    {operationalPoints.map((op) => (
+    {waypoints.map((waypoint) => (
       <div
-        key={op.id}
```
Comment (emersion) [inline ui-manchette/src/components/OperationalPointList.tsx:18]: Yay, this no longer causes duplicate `key` attributes if the same OP appears twice in the list!

**Review by emersion — APPROVED** [review]
LGTM, very nice cleanup 🙇

File: ui-manchette/src/types.ts (hunk 2 above)
Comment (emersion) [reply ui-manchette/src/types.ts:4]: Cool, something to keep in mind when writing the OSRD patch to upgrade to this new API.

File: ui-manchette-with-spacetimechart/src/hooks/useManchetteWithSpaceTimeChart.ts [hunk 5]
```diff
@@ -74,10 +71,10 @@ const useManchettesWithSpaceTimeChart = (
 
   // Memoize computedOperationalPoints to avoid recalculations unless the dependencies change
   const operationalPointsToDisplay = useMemo(
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/hooks/useManchetteWithSpaceTimeChart.ts]:
```suggestion
  const waypointsToDisplay = useMemo(
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 6]
```diff
@@ -10,28 +10,29 @@ import type {
 
 import { BASE_OP_HEIGHT, MAX_TIME_WINDOW, MAX_ZOOM_X, MIN_ZOOM_X } from './consts';
 import { calcTotalDistance, getHeightWithoutLastWaypoint, msToS } from './utils';
-export type OperationalPointsOptions = { isProportional: boolean; yZoom: number; height: number };
+
+type OperationalPointsOptions = { isProportional: boolean; yZoom: number; height: number };
 
 export const calcOperationalPointsToDisplay = (
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
export const calcWaypointsToDisplay = (
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 7]
```diff
@@ -10,28 +10,29 @@ import type {
 
 import { BASE_OP_HEIGHT, MAX_TIME_WINDOW, MAX_ZOOM_X, MIN_ZOOM_X } from './consts';
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]: We should rename it as well

```suggestion
import { BASE_WAYPOINT_HEIGHT, MAX_TIME_WINDOW, MAX_ZOOM_X, MIN_ZOOM_X } from './consts';
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 8]
```diff
@@ -10,28 +10,29 @@ import type {
...
 import { calcTotalDistance, getHeightWithoutLastWaypoint, msToS } from './utils';
-export type OperationalPointsOptions = { isProportional: boolean; yZoom: number; height: number };
+
+type OperationalPointsOptions = { isProportional: boolean; yZoom: number; height: number };
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
type WaypointsOptions = { isProportional: boolean; yZoom: number; height: number };
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 9]
```diff
@@ -10,28 +10,29 @@ import type {
...
 
 export const calcOperationalPointsToDisplay = (
-  operationalPoints: OperationalPointType[],
+  waypoints: Waypoint[],
   { height, isProportional, yZoom }: OperationalPointsOptions
-): StyledOperationalPointType[] => {
-  if (!isProportional || operationalPoints.length === 0) {
+): InteractiveWaypoint[] => {
+  if (!isProportional || waypoints.length === 0) {
     // For non-proportional display, we always display all the operational points:
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
    // For non-proportional display, we always display all the waypoints:
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 10]
```diff
@@ -10,28 +10,29 @@ import type {
...
   }
 
   // For proportional display, we only display points that do not overlap with
   // the last displayed point:
-  const result: StyledOperationalPointType[] = [{ ...operationalPoints[0], display: true }];
-  const totalDistance = calcTotalDistance(operationalPoints);
+  const result: InteractiveWaypoint[] = [{ ...waypoints[0], display: true }];
+  const totalDistance = calcTotalDistance(waypoints);
   const heightWithoutFinalOp = getHeightWithoutLastWaypoint(height);
   let lastDisplayedOP = result[0];
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
  let lastDisplayedWaypoint = result[0];
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 11]
```diff
@@ -10,28 +10,29 @@ import type {
...
+    return waypoints.map((waypoint) => ({ ...waypoint, display: true }));
   }
 
   // For proportional display, we only display points that do not overlap with
   // the last displayed point:
-  const result: StyledOperationalPointType[] = [{ ...operationalPoints[0], display: true }];
-  const totalDistance = calcTotalDistance(operationalPoints);
+  const result: InteractiveWaypoint[] = [{ ...waypoints[0], display: true }];
+  const totalDistance = calcTotalDistance(waypoints);
   const heightWithoutFinalOp = getHeightWithoutLastWaypoint(height);
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
  const heightWithoutFinalWaypoint = getHeightWithoutLastWaypoint(height);
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 12]
```diff
@@ -10,28 +10,29 @@ import type {
...
   let lastDisplayedOP = result[0];
 
   // We iterate through all points, and only add them if they don't collide
   // with the last visible point:
-  for (let i = 1; i < operationalPoints.length; i++) {
-    const op = operationalPoints[i];
+  for (let i = 1; i < waypoints.length; i++) {
+    const op = waypoints[i];
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
    const waypoint = waypoints[i];
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 13]
```diff
@@ -56,15 +57,15 @@ export const calcOperationalPointsToDisplay = (
...
 ) => {
-  if (operationalPoints.length < 2) return [];
-  const totalDistance = calcTotalDistance(operationalPoints);
+  if (waypoints.length < 2) return [];
+  const totalDistance = calcTotalDistance(waypoints);
   const heightWithoutFinalOp = getHeightWithoutLastWaypoint(height);
 
-  return operationalPoints.map((op, index) => {
-    const nextOp = operationalPoints.at(index + 1);
+  return waypoints.map((op, index) => {
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
  return waypoints.map((waypoint, index) => {
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 14]
```diff
@@ -56,15 +57,15 @@ export const calcOperationalPointsToDisplay = (
...
+    const nextOp = waypoints.at(index + 1);
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
    const nextWaypoint = waypoints.at(index + 1);
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 15]
```diff
@@ -104,9 +105,9 @@ export const computeTimeWindow = (trains: ProjectPathTrainResult[]) => {
 };
 
 export const getOperationalPointsWithPosition = (
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts]:
```suggestion
export const getWaypointsWithPosition = (
```

File: ui-manchette-with-spacetimechart/src/helpers.ts [hunk 16]
```diff
@@ -1,7 +1,7 @@
 import type {
-  OperationalPointType,
+  InteractiveWaypoint,
   ProjectPathTrainResult,
-  StyledOperationalPointType,
+  Waypoint,
 } from '@osrd-project/ui-manchette/dist/types';
 import type {
   OperationalPoint,
```
Comment (SharglutDev) [inline ui-manchette-with-spacetimechart/src/helpers.ts:7]: Concerns the spacetime part but this should be renamed too ?

File: ui-manchette/src/components/Manchette.tsx [hunk 17]
```diff
@@ -5,10 +5,10 @@ import cx from 'classnames';
 
 import { INITIAL_OP_LIST_HEIGHT, MAX_ZOOM_Y, MIN_ZOOM_Y } from './consts';
 import OperationalPointList from './OperationalPointList';
```
Comment (SharglutDev) [inline ui-manchette/src/components/Manchette.tsx]: The components should be renamed too

File: ui-manchette/src/components/OperationalPoint.tsx [hunk 18]
```diff
@@ -32,9 +32,9 @@ const OperationalPoint = ({
     >
       <div className="op-position justify-self-start text-end">{positionMmToKm(position)}</div>
```
Comment (SharglutDev) [inline ui-manchette/src/components/OperationalPoint.tsx]: class should be renamed too

```suggestion
      <div className="waypoint-position justify-self-start text-end">{positionMmToKm(position)}</div>
```

File: ui-manchette/src/stories/OperationalPoint.stories.ts
Comment (SharglutDev) [inline ui-manchette/src/stories/OperationalPoint.stories.ts:1]: Stories file name should be renamed too

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for this ! Missed some remaining `op` / `operationalPoints`, I might have missed some

**Comment by clarani** [comment]
Yes I wanted to rename the files/components/styles in an other commit :) But I can do it now

File: ui-manchette-with-spacetimechart/src/helpers.ts (hunk 16 above)
Comment (clarani) [reply ui-manchette-with-spacetimechart/src/helpers.ts:7]: I think we should rename it yes, but maybe later ?

**Comment by clarani** [comment]
I renamed all the occurences of operationalPoint and op, except the ones which were related to the spaceTimeChart :)

File: ui-manchette-with-spacetimechart/src/helpers.ts (hunk 16 above)
Comment (SharglutDev) [reply ui-manchette-with-spacetimechart/src/helpers.ts:7]: Sure

**Review by SharglutDev — APPROVED** [review]
Lgtm and tested, thank you for this great refacto :)

We should create a ticket to do the same in spacetime and speedspace

**Pull request merged**